### ISSUE TITLE
The onnx model(which is transfor by export_onnx.py) out put is differ from pytoch model

### ISSUE THREAD

**Issue by Genlk** [RangiLyu/nanodet#478]
def image_preprocess(img_path):
    img = cv2.imread(img_path).astype("float32")/255
    # mean = [103.53, 116.28, 123.675] # Image net values
    # std = [57.375, 57.12, 58.395]
    mean = [113.533554, 118.14172, 123.63607]
    std = [21.405144, 21.405144, 21.405144]
    mean = np.array(mean, dtype=np.float32).reshape(1, 1, 3) / 255
    std = np.array(std, dtype=np.float32).reshape(1, 1, 3) / 255
    img = (img - mean) / std
    img = np.transpose(img, (2, 0, 1))
    img = np.expand_dims(img, axis=0)
    return img

def test_onnx_model(onnx_model,img_path=None):
    if img_path is None:
        img_path = "path for img"
    imgdata = image_preprocess(img_path)
    sess = rt.InferenceSession(onnx_model)
    input_name = sess.get_inputs()[0].name
    output_detect_name = sess.get_outputs()[0].name
    pred_onnx0= sess.run([output_detect_name], {input_name: imgdata})
    print("outputs:")
    print(np.array(pred_onnx0))

**Comment by deepak046** [comment]
I had the same issue now, and there's a _forward_onnx class function inside nanodet/model/head/nanodet_plus_head.py which calculates the sigmoid of the class predictions whereas the pytorch model's forward function doesn't do it. That's why the output differs.

def _forward_onnx(self, feats):
    """only used for onnx export"""
    outputs = []
    for feat, cls_convs, gfl_cls in zip(
        feats,
        self.cls_convs,
        self.gfl_cls,
    ):
        for conv in cls_convs:
            feat = conv(feat)
        output = gfl_cls(feat)
        outputs.append(output.flatten(start_dim=2))
    return torch.cat(outputs, dim=2).permute(0, 2, 1)

I changed the function to this and the outputs match now. Make sure you also do the pre-processing and post-processing for the onnx exported model exactly the same way as it's being done for the pytorch model.
